lib/phoenix_component.ex

```diff
@@ -1025,13 +1221,146 @@ defmodule Phoenix.Component do
+        compile_error!(first_slot_line, env.file, """
+        slots must be defined before the first function clause at line #{meta[:line]}
+        """)
+      end
+    end
+  end
+
+  defmacro slot(name, opts \\ []) when is_atom(name) and is_list(opts) do
+    {block, opts} = Keyword.pop(opts, :do, nil)
+
+    quote do
+      Phoenix.Component.__slot__!(
+        __MODULE__,
+        unquote(name),
+        unquote(opts),
+        __ENV__.line,
+        __ENV__.file,
+        fn -> unquote(block) end
+      )
+    end
+  end
+
+  defmacro slot(name, opts, do: block) when is_atom(name) and is_list(opts) do
+    quote do
+      Phoenix.Component.__slot__!(
+        __MODULE__,
+        unquote(name),
+        unquote(opts),
+        __ENV__.line,
+        __ENV__.file,
+        fn -> unquote(block) end
+      )
+    end
+  end
+
+  @doc false
+  def __slot__!(module, name, opts, line, file, block_fun \\ fn -> :ok end) do
+    {doc, opts} = Keyword.pop(opts, :doc, nil)
+
+    unless is_binary(doc) or is_nil(doc) or doc == false do
+      compile_error!(line, file, ":doc must be a string or false, got: #{inspect(doc)}")
+    end
+
+    {required, opts} = Keyword.pop(opts, :required, false)
+
+    unless is_boolean(required) do
+      compile_error!(line, file, ":required must be a boolean, got: #{inspect(required)}")
+    end
+
+    Module.put_attribute(module, :__slot__, name)
+
+    try do
+      block_fun.()
+    rescue
+      e ->
+        Module.put_attribute(module, :__slot__, nil)
+        Module.delete_attribute(module, :__slot_attrs__)
+        Module.register_attribute(module, :__slot_attrs__, accumulate: true)
+        reraise e, __STACKTRACE__
+    end
+
+    slot_attrs =
+      module
+      |> Module.get_attribute(:__slot_attrs__)
+      |> Enum.reverse()
+
+    slot = %{
+      name: name,
+      required: required,
+      opts: opts,
+      doc: doc,
+      line: line,
+      attrs: slot_attrs
+    }
+
+    validate_slot!(module, slot, line, file)
+
+    Module.put_attribute(module, :__slots__, slot)
+    Module.put_attribute(module, :__slot__, nil)
+    Module.delete_attribute(module, :__slot_attrs__)
+    Module.register_attribute(module, :__slot_attrs__, accumulate: true)
+  end
+
+  defp validate_slot!(module, slot, line, file) do
+    slots = get_slots(module)
+
+    if Enum.find(slots, &(&1.name == slot.name)) do
+      compile_error!(line, file, """
+      a duplicate slot with name #{inspect(slot.name)} already exists
+      """)
+    end
+
+    if slot.name == :inner_block and not Enum.empty?(slot.attrs) do
+      compile_error!(line, file, """
+      cannot define attributes in slot #{inspect(slot.name)}
+      """)
+    end
+
+    for {attr_name, attr_defs} <- Enum.group_by(slot.attrs, & &1.name),
+        length(attr_defs) > 1,
+        %{line: line} <- attr_defs do
+      compile_error!(line, file, """
+      a duplicate attribute with name #{inspect(attr_name)} in slot #{inspect(slot.name)} already exists
+      """)
```

Comment:
IMO we should check this inside `validate_attr_type`. The main reason is that the stacktrace will point precisely to the duplicate attribute declaration. Raising here will point to the slot definition.